src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java

```diff
@@ -213,92 +219,236 @@ public CassandraStorage(
-      int currentVersion = database.getVersion();
-      MigrationRepository migrationRepo = new MigrationRepository("db/cassandra");
-      if (currentVersion < migrationRepo.getLatestVersion()) {
-        LOG.warn("Starting db migration from {} to {}…", currentVersion, migrationRepo.getLatestVersion());
-
-        if (4 <= currentVersion) {
-          List<String> otherRunningReapers = session.execute("SELECT reaper_instance_host FROM running_reapers").all()
-              .stream()
-              .map((row) -> row.getString("reaper_instance_host"))
-              .filter((reaperInstanceHost) -> !AppContext.REAPER_INSTANCE_ADDRESS.equals(reaperInstanceHost))
-              .collect(Collectors.toList());
-
-          Preconditions.checkState(
-              otherRunningReapers.isEmpty(),
-              "Database migration can not happen with other reaper instances running. Found ",
-              StringUtils.join(otherRunningReapers));
+      while (!migrationRequired.isPresent()) {
+        int currentVersion = database.getVersion();
+        MigrationRepository migrationRepo = new MigrationRepository("db/cassandra");
+        if (currentVersion < migrationRepo.getLatestVersion()) {
+          if (lockSchemaMigration(session, concurrentMigrationsAreEnabled, version)) {
+            // We own the lock and a migration is required
+            migrationRequired = Optional.of(true);
+          } else {
+            // Migration is required but another reaper instance owns the lock
+            LOG.info("Cannot grab the lock on schema migration. Waiting for it to be released...");
+            try {
+              Thread.sleep(10000);
+            } catch (InterruptedException e) {
+              // TODO Auto-generated catch block
+              throw new ReaperException(e);
+            }
+          }
+        } else {
+          LOG.info("Schema is up to date, skipping migration.");
+          migrationRequired = Optional.of(false);
         }
+      }
 
-        migrate(database.getVersion(), migrationRepo, session);
-        // some migration steps depend on the Cassandra version, so must be rerun every startup
-        Migration016.migrate(session, keyspace);
-        // Switch metrics table to TWCS if possible, this is intentionally executed every startup
-        Migration021.migrate(session, keyspace);
-      } else {
-        LOG.info(
-            String.format("Keyspace %s already at schema version %d", session.getLoggedKeyspace(), currentVersion));
+      if (migrationRequired.get() == true) {
+        int currentVersion = database.getVersion();
+        MigrationRepository migrationRepo = new MigrationRepository("db/cassandra");
+        if (currentVersion < migrationRepo.getLatestVersion()) {
+          if (isInSidecarMode && !concurrentMigrationsAreEnabled) {
+            throw new ReaperException(
+                "Running in sidecar mode and concurrent migrations are not enabled.");
```

Comment:
trivial: i doubt this line was over 120 characters and needed to be wrapped? ide configured wrong?